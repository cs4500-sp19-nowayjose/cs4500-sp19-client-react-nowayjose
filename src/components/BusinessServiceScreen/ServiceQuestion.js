import React from 'react'

function BooleanFilter({title, description, answer, setAnswer}) {
  return (
    <div className="provider-filter boolean-filter">
      <h4>{title}</h4>
      <div>{description}</div>
      <input type="checkbox" value={answer || 0}
        onChange={(e) => setAnswer(e.target.checked)} />
    </div>
  )
}

function MinMaxFilter({title, description, answer, setAnswer}) {
  let update = (key, value) => {
    let v = {}
    v[key] = answer
    setAnswer(Object.assign(answer || {min: 0, max: 0}, v))
  }
  return (
    <div className="provider-filter range-filter">
      <h4>{title}</h4>
      <div>{description}</div>
      <label>Min</label>
      <input type="number" value={answer}
        onChange={(e) => update("min", e.target.value)} />
      <label>Max</label>
      <input type="number" value={answer}
        onChange={(e) => update("max", e.target.value)} />
    </div>
  )
}

function MultipleChoiceFilter({title, description, choiceOptions, answer, setAnswer}) {
  return (
    <div className="provider-filter choice-filter">
      <h4>{title}</h4>
      <div>{description}</div>
      <select value={answer || 0} onChange={(e) => setAnswer(parseInt(e.target.value))}>
        {
          choiceOptions.map((option, i) => <option value={i}>{option}</option>)
        }
      </select>
    </div>
  )
}

export default function ServiceQuestion({question, answer, setAnswer}) {
  if (question.serviceQuestionType === "YESORNO") {
    return (
      <BooleanFilter
        title={question.title}
        description={question.description}
        answer={answer}
        setAnswer={setAnswer} />
    )
  } else if (question.serviceQuestionType === "MINMAX") {
    return (
      <MinMaxFilter
        title={question.title}
        description={question.description}
        answer={answer}
        setAnswer={setAnswer} />
    )
  } else if (question.serviceQuestionType === "MULTIPLECHOICES") {
    return (
      <MultipleChoiceFilter
        title={question.title}
        description={question.description}
        choiceOptions={question.choiceOptions}
        answer={answer}
        setAnswer={setAnswer} />
    )
  } else {
    return null;
  }
}
