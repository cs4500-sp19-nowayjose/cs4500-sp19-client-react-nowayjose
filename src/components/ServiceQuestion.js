import React from 'react'
import ServiceQuestionService from '../services/ServiceQuestionService'
class ServiceQuestions extends React.Component {
    constructor(props) {
        super(props)
        this.serviceQuestionService = ServiceQuestionService.getInstance()
        this.state = {
            serviceQuestions: [{
              id: 1288345,
              title: "Hi there"
            }]
        }
    }
    componentDidMount() {
        // this.serviceQuestionService
        //     .findAllServiceQuestions()
        //     .then(serviceQuestions =>
        //         this.setState({
        //             serviceQuestions: serviceQuestions
        //         })
        //     )
    }

    deleteQuestion(id) {
      this.serviceQuestionService.delete(id)
      this.setState(Object.assign(this.state, {
        serviceQuestions: this.state.serviceQuestions.filter(question => question.id != id)
      }))
    }

    render() {
        return(
            <div>
                <h3>Service Questions</h3>
                <table className="table">
                    <tbody>
                    {
                        this.state.serviceQuestions
                            .map(serviceQuestion =>
                                <tr key={serviceQuestion.id}>
                                    <td>
                                    to={`/admin/service-questions/${serviceQuestion.id}`}>
                                    {serviceQuestion.title}</td>
                                  <span onClick={() => this.deleteQuestion(serviceQuestion.id)}>X</span>
                                </tr>
                            )
                    }
                    </tbody>
                </table>
            </div>
        )
    }
}

export default ServiceQuestions
