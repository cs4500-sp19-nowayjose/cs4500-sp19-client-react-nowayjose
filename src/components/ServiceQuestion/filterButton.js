import React from 'react'
import { View } from 'react';

const FilterButton = ({ onClickFilterData }) => {
  return (
    <button
      onClick={onClickFilterData}
      testID="search-btn-filter-service-question"
      className="btn-lg btn"
      style={{ position: 'fixed', right: 10, color: '#FEC107'}}
    >
      <i className="fas fa-search fa-2x"></i>
    </button>
  )
}

export default FilterButton;
