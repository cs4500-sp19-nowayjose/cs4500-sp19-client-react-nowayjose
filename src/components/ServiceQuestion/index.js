import React from 'react'
import ServiceQuestionService from '../../services/ServiceQuestionService'
import FilterButton from './filterButton'
import QuestionTable from './questionTable'
import PaginationTool from './paginationTool'

class ServiceQuestions extends React.Component {
  serviceQuestionService = ServiceQuestionService.getInstance()
  state = {
    serviceQuestions: [],
    page: 0,
    resultsPerPage: 10,
    filter: {
      title: '',
      description: '',
    }
  }


  componentDidMount() {
    this.serviceQuestionService
      .findAllServiceQuestions()
      .then(serviceQuestions => {
        if (serviceQuestions.status === 500) {
          return;
        }
        this.setState({ serviceQuestions });
      })
  }


  incrPage = (incr) => {
    this.setPage(this.state.page + incr)
  }

  setPage = (pageNumber) => {
    if (pageNumber >= 0 && pageNumber < Math.ceil(this.state.serviceQuestions.length / this.state.resultsPerPage)) {
      this.setState(Object.assign(this.state, {
        page: pageNumber
      }))
    }
  }

  setResultsPerPage = (e) => {
    const value = e.target.value;
    let resultsPerPage = this.state.serviceQuestions.length
    if (value !== "all") {
      resultsPerPage = parseInt(value)
    }
    this.setState(Object.assign(this.state, {
      resultsPerPage: resultsPerPage
    }))
  }

  deleteQuestion = (id) => {
    this.serviceQuestionService.delete(id)
    this.setState(Object.assign(this.state, {
      serviceQuestions: this.state.serviceQuestions.filter(question => question.id !== id)
    }))
  }

  handleFilterChange = (event, type) => {
    const { value } = event.target;
    this.setState(prevState => {
      return {
        filter: {
          ...prevState.filter,
          [type]: value,
        }
      }
    });
  }

  filterData = async () => {
    const result = await this.serviceQuestionService.findServiceQuestionByCriteria(this.state.filter);
    this.setState({
      serviceQuestions: result.length ? result : [],
      page: 0,
      resultsPerPage: 10,
      filter: { title: '', description: '' },
    })
  }

  render() {
    const { filter, page, resultsPerPage, serviceQuestions } = this.state;
    return (
      <div>
        <h3>Service Questions</h3>
        <QuestionTable 
          filterState={filter}
          pageState={{ page, resultsPerPage }}
          handleFilterChange={this.handleFilterChange}
          deleteQuestion={this.deleteQuestion}
          serviceQuestions={serviceQuestions}
        />
        <div>
          <PaginationTool
            incrPage={this.incrPage}
            setResultsPerPage={this.setResultsPerPage}
            setPage={this.setPage}
            paginationNumbers={[...Array(Math.ceil(serviceQuestions.length / resultsPerPage)).keys()]}
          />
          <FilterButton onClickFilterData={this.filterData} />
        </div>
      </div>
    )
  }
}

export default ServiceQuestions
