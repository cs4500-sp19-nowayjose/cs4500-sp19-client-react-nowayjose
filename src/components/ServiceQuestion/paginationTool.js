import React from 'react'

export default function paginationTool({
  incrPage,
  setResultsPerPage,
  setPage,
  paginationNumbers,
}) {
  return (
    <React.Fragment>
      <span testID="prev-button" onClick={() => incrPage(-1)}>Prev</span>
      <Pagination
        testID="pagination-buttons"
        paginationNumbers={paginationNumbers}
        setPage={setPage}
      />
      <span testID="next-button" onClick={() => incrPage(1)}>Next</span>
      <select onChange={setResultsPerPage}>
        <option value="10">10</option>
        <option value="25">25</option>
        <option value="50">50</option>
        <option value="100">100</option>
        <option value="all">all</option>
      </select>
    </React.Fragment>
  )
}

function Pagination({ paginationNumbers, setPage }) {
  return paginationNumbers.map(pageNumber =>
    <span 
      key={`${pageNumber}pageNumberQuestionService`} 
      onClick={() => setPage(pageNumber)}
    >
      {pageNumber + 1}
    </span>
  )
}
