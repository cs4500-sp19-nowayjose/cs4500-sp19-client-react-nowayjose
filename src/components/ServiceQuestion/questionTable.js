import React from 'react'

export default function questionTable({ 
  filterState,
  pageState,
  handleFilterChange,
  serviceQuestions,
  deleteQuestion
}) {
  const { page, resultsPerPage } = pageState;
  return (
    <table className="table">
      <TableHeader filterState={filterState} handleFilterChange={handleFilterChange} />
      <TableBody
        serviceQuestions={serviceQuestions}
        page={page}
        resultsPerPage={resultsPerPage}
        deleteQuestion={deleteQuestion}
      />
    </table>
  )
}

function TableHeader({ filterState, handleFilterChange }) {
  const { title, description, serviceQuestionType } = filterState;
  return (
    <thead>
      <TitleHeader />
      <FilterHeader
        title={title}
        description={description}
        serviceQuestionType={serviceQuestionType}
        onFilterChange={handleFilterChange}
      />
    </thead>
  )
}

function TitleHeader() {
  return (
    <tr>
      <th>Title</th>
      <th>Description</th>
      <th>Type</th>
      <th> </th>
    </tr>
  )
}

function FilterHeader({ title, description, serviceQuestionType, onFilterChange }) {
  return (
    <tr>
      <th><input type="text" value={title} onChange={(e) => onFilterChange(e, 'title')}/></th>
      <th><input type="text" value={description} onChange={(e) => onFilterChange(e, 'description')}/></th>
      <th><input type="text" value={serviceQuestionType} disabled/></th>
      <th>
        <button type="button" className="btn btn-primary">Add</button>
      </th>
      <th>
        <button type="button" className="btn btn-success">Save</button>
      </th>
    </tr>
  )
}

function TableBody({ serviceQuestions, page, resultsPerPage, deleteQuestion }) {
  return (
    <tbody>{
      serviceQuestions
        .slice(page * resultsPerPage, (page + 1) * resultsPerPage)
        .map(serviceQuestion => <TableRow serviceQuestion={serviceQuestion} deleteQuestion={deleteQuestion} />)
    }
    </tbody>
  )
}

function TableRow({ serviceQuestion, deleteQuestion }) {
  const { id,  title, description, serviceQuestionType } = serviceQuestion
  return (
    <tr key={id}>
      <td>
        to={`/admin/service-questions/${id}`}>
        {title}</td>
      <td>{description}</td>
      <td>{serviceQuestionType}</td>
      <span onClick={() => deleteQuestion(id)}>X</span>
    </tr>
  )
}