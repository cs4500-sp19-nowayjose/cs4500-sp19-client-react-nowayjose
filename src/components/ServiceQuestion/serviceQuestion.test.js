
import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';

import ServiceQuestion from './'
import setUpService from '../../services/ServiceQuestion.mock'
import serviceQuestionMockData from '../../data/serviceQuestions.mock.json'

describe('service question on admin panel rendering test', function() {
  let ServiceQuestionComponent
  beforeAll(function() {
    setUpService();
    ServiceQuestionComponent = renderer.create(
      <Router>
        <ServiceQuestion
          isTest
          testData={serviceQuestionMockData}
        />
      </Router>
    )
  })

  test('Render service questions table', function() {
    const component = ServiceQuestionComponent;
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Filtering title works', async function() {
    const component = ServiceQuestionComponent;
    const titleFieldFilter = component.root.findByProps({ testID: 'service-question-title-filter' })
    const filterButton = component.root.findByProps({ testID: "search-btn-filter-service-question" })
    await titleFieldFilter.props.onChange({ target: { value: 'houses' }})
    expect(titleFieldFilter.props.value).toBe("houses")
    await filterButton.props.onClick()
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
