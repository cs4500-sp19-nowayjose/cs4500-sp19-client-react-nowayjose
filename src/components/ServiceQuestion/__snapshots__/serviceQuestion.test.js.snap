// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`service question on admin panel rendering test Filtering title works 1`] = `
<div>
  <h3>
    Service Questions
  </h3>
  <table
    className="table"
  >
    <thead>
      <tr>
        <th>
          Title
        </th>
        <th>
          Description
        </th>
        <th>
          Type
        </th>
        <th>
           
        </th>
      </tr>
      <tr>
        <th>
          <input
            onChange={[Function]}
            testID="service-question-title-filter"
            type="text"
            value=""
          />
        </th>
        <th>
          <input
            onChange={[Function]}
            type="text"
            value=""
          />
        </th>
        <th>
          <input
            disabled={true}
            type="text"
          />
        </th>
        <th>
          <button
            className="btn btn-primary"
            type="button"
          >
            Add
          </button>
        </th>
        <th>
          <button
            className="btn btn-success"
            type="button"
          >
            Save
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        testID="service_question_row_1"
      >
        <td>
          <a
            href="/admin/service-questions/1"
            onClick={[Function]}
          >
            What size houses can you clean (in square feet)?
          </a>
        </td>
        <td>
          Select the minimum and maximum size of houses that you can clean.
        </td>
        <td />
        <span
          onClick={[Function]}
        >
          X
        </span>
      </tr>
    </tbody>
  </table>
  <div>
    <span
      onClick={[Function]}
    >
      Prev
    </span>
    <span
      onClick={[Function]}
    >
      1
    </span>
    <span
      onClick={[Function]}
    >
      Next
    </span>
    <select
      onChange={[Function]}
    >
      <option
        value="10"
      >
        10
      </option>
      <option
        value="25"
      >
        25
      </option>
      <option
        value="50"
      >
        50
      </option>
      <option
        value="100"
      >
        100
      </option>
      <option
        value="all"
      >
        all
      </option>
    </select>
    <button
      className="btn-lg btn"
      onClick={[Function]}
      style={
        Object {
          "color": "#FEC107",
          "position": "fixed",
          "right": 10,
        }
      }
      testID="search-btn-filter-service-question"
    >
      <i
        className="fas fa-search fa-2x"
      />
    </button>
  </div>
</div>
`;

exports[`service question on admin panel rendering test Render service questions table 1`] = `
<div>
  <h3>
    Service Questions
  </h3>
  <table
    className="table"
  >
    <thead>
      <tr>
        <th>
          Title
        </th>
        <th>
          Description
        </th>
        <th>
          Type
        </th>
        <th>
           
        </th>
      </tr>
      <tr>
        <th>
          <input
            onChange={[Function]}
            testID="service-question-title-filter"
            type="text"
            value=""
          />
        </th>
        <th>
          <input
            onChange={[Function]}
            type="text"
            value=""
          />
        </th>
        <th>
          <input
            disabled={true}
            type="text"
          />
        </th>
        <th>
          <button
            className="btn btn-primary"
            type="button"
          >
            Add
          </button>
        </th>
        <th>
          <button
            className="btn btn-success"
            type="button"
          >
            Save
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        testID="service_question_row_1"
      >
        <td>
          <a
            href="/admin/service-questions/1"
            onClick={[Function]}
          >
            What size houses can you clean (in square feet)?
          </a>
        </td>
        <td>
          Select the minimum and maximum size of houses that you can clean.
        </td>
        <td />
        <span
          onClick={[Function]}
        >
          X
        </span>
      </tr>
      <tr
        testID="service_question_row_4"
      >
        <td>
          <a
            href="/admin/service-questions/4"
            onClick={[Function]}
          >
            Do you have any pet?
          </a>
        </td>
        <td>
          Select how many pet you have
        </td>
        <td />
        <span
          onClick={[Function]}
        >
          X
        </span>
      </tr>
    </tbody>
  </table>
  <div>
    <span
      onClick={[Function]}
    >
      Prev
    </span>
    <span
      onClick={[Function]}
    >
      1
    </span>
    <span
      onClick={[Function]}
    >
      Next
    </span>
    <select
      onChange={[Function]}
    >
      <option
        value="10"
      >
        10
      </option>
      <option
        value="25"
      >
        25
      </option>
      <option
        value="50"
      >
        50
      </option>
      <option
        value="100"
      >
        100
      </option>
      <option
        value="all"
      >
        all
      </option>
    </select>
    <button
      className="btn-lg btn"
      onClick={[Function]}
      style={
        Object {
          "color": "#FEC107",
          "position": "fixed",
          "right": 10,
        }
      }
      testID="search-btn-filter-service-question"
    >
      <i
        className="fas fa-search fa-2x"
      />
    </button>
  </div>
</div>
`;
